Sources/GRPC/HTTP2ToRawGRPCServerCodec.swift

```diff
@@ -335,15 +326,21 @@ internal final class HTTP2ToRawGRPCServerCodec: ChannelDuplexHandler, GRPCServer
       // Always end stream for status and trailers.
       let payload = HTTP2Frame.FramePayload.headers(.init(headers: trailers, endStream: true))
       self.context.write(self.wrapOutboundOut(payload), promise: promise)
-
-      if self.isReading {
-        self.flushPending = true
-      } else {
-        self.context.flush()
-      }
+      // We'll always flush on end.
+      self.markFlushPoint()
 
     case let .failure(error):
       promise?.fail(error)
     }
   }
+
+  /// Mark a flush as pending - to be emitted once the read completes - if we're currently reading,
+  /// or emit a flush now if we are not.
+  private func markFlushPoint() {
+    if self.isReading {
+      self.flushPending = true
+    } else {
+      self.context.flush()
```

Comment:
Don't we need to unset `flushPending` here?

Reply:
I don't think it's strictly necessary; we'll only set it when we're reading and we'll unset it before we flush in channel read complete. That said, that seems a little brittle and there's no harm in unsetting it here too.

Reply:
For that matter, do we do anything with `flushPending` at all?

Reply:
Oh never mind, I'm clearly looking in the wrong place.